.github/workflows/main.yml

```diff
@@ -14,23 +14,70 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        docker-image:  # See https://hub.docker.com/_/python/tags for images
-          - python:2.7.18-alpine3.11
-          - python:3.11.4-alpine3.18
-    container: ${{ matrix.docker-image }}
+        python-version: ['2.7', '3.11']
     steps:
       - name: Checkout code
         uses: actions/checkout@v3
+      - uses: LizardByte/setup-python-action@master
+        with:
+          python-version: ${{matrix.python-version}}
+
+      - uses: actions/cache@v3
+        name: Setup cache for running pre-commit fast
+        with:
+            path: ~/.cache/pre-commit
+            key: pre-commit|${{ env.pythonLocation }}|${{ hashFiles('.pre-commit-config.yaml') }}
+
+      - run: echo "::add-matcher::.github/workflows/python-warning-matcher.json"
+        name: "Setup GitHub for reporting Python warnings as annotations in pull request code review"
+
+      - uses: pre-commit/action@v3.0.0
+        name: Run pre-commit checks (no spaces at end of lines, etc)
+        if: ${{ matrix.python-version != '2.7' }}
+        with:
+            extra_args: --all-files --verbose
+        env:
+          PYTHONDEVMODE: yes
+          SKIP: no-commit-to-branch
 
       - name: Install python 2 dependencies
-        if: ${{ startsWith(matrix.docker-image, 'python:2.7.18') }}
+        if: ${{ matrix.python-version == '2.7' }}
         run: pip install enum
 
-      - name: Install dependencies
-        run: pip install mock pytest
+      - name: Install common dependencies for Python${{matrix.python-version}}
+        run: pip install mock pytest-coverage pytest-mock
+
+      - name: Run Pytest tests for Python${{matrix.python-version}}
+        run: >
+          pytest --cov scripts scripts/ -vv -rA
+          --junitxml=.git/pytest${{matrix.python-version}}.xml
+          --cov-report term-missing
+          --cov-report html:.git/coverage${{matrix.python-version}}.html
+          --cov-report xml:.git/coverage${{matrix.python-version}}.xml
```

Comment:
Why do both of these need to be reported? Can't it just the html be reported? (or the xml)

Reply:
We could save the .html as artifact, but I removed it now since we have Covecov working soon.